pyproject.toml

```diff
@@ -1,57 +1,135 @@
+  "azure-mgmt-rdbms~=10.1.0",
+  "azure-mgmt-resource~=20.1.0",
+  "azure-mgmt-storage~=19.1.0",
+  "azure-storage-blob~=12.15.0",
+  "azure-storage-file-datalake~=12.10.0",
+  "azure-storage-file-share~=12.10.0",
+  "chevron~=0.14.0",
+  "cryptography~=3.4.0",
+  "dnspython~=2.3.0",
+  "dotmap~=1.3.0",
+  "msal~=1.21.0",
+  "psycopg2~=2.9.0",
+  "pulumi~=3.67.0",
+  "pulumi-azure-native~=1.95.0",
+  "pytz~=2022.7.0",
+  "PyYAML~=6.0",
+  "rich~=13.4.2",
+  "simple-acme-dns~=1.2.0",
+  "typer~=0.9.0",
+  "websocket-client~=1.5.0",
+]
+
+[project.scripts]
+dsh = "data_safe_haven.cli:main"
 
-[tool.poetry.dependencies]
-python = "^3.10"
-azure-core = "^1.26.0"
-azure-identity = "^1.12.0"
-azure-keyvault-certificates = "^4.6.0"
-azure-keyvault-keys = "^4.6.0"
-azure-keyvault-secrets = "^4.6.0"
-azure-mgmt-automation = "^1.0.0"
-azure-mgmt-compute = "^29.1.0"
-azure-mgmt-containerinstance = "^9.2.0"
-azure-mgmt-dns = "^8.0.0"
-azure-mgmt-keyvault = "^9.3.0"
-azure-mgmt-msi = "^6.1.0"
-azure-mgmt-network = "^20.0.0"
-azure-mgmt-rdbms = "^10.1.0"
-azure-mgmt-resource = "^20.1.0"
-azure-mgmt-storage = "^19.1.0"
-azure-storage-blob = "^12.14.0"
-azure-storage-file-datalake = "^12.10.0"
-azure-storage-file-share = "^12.10.0"
-chevron = "^0.14.0"
-cryptography = "^3.4.0"
-dnspython = "^2.3.0"
-dotmap = "^1.3.0"
-msal = "^1.21.0"
-psycopg2 = "^2.9.0"
-pulumi = "^3.67.0"
-pulumi-azure-native = "^1.95.0"
-pytz = "^2022.7.0"
-PyYAML = "^6.0"
-rich = "^13.4.2"
-simple-acme-dns = "^1.2.0"
-typer = "^0.9.0"
-websocket-client = "^1.5.0"
+[tool.hatch.version]
+path = "data_safe_haven/__about__.py"
 
-[tool.poetry.dev-dependencies]
-black = "^22.12.0"
-mypy = "^1.0.0"
-pytest = "^7.2"
-types-chevron = "^0.14.0"
-types-psycopg2 = "^2.9.0"
-types-pytz = "^2022.7.0"
-types-PyYAML = "^6.0"
-types-requests = "^2.28.0"
-types-setuptools = "^67.3.0"
+[tool.hatch.envs.lint]
+detached = true
+dependencies = [
+  "black>=23.1.0",
+  "mypy>=1.0.0",
+  "ruff>=0.0.243",
+]
 
-[tool.poetry.scripts]
-dsh = "data_safe_haven.cli:main"
+[tool.hatch.envs.lint.scripts]
+typing = "mypy --strict --install-types --non-interactive --namespace-packages --explicit-package-bases {args:data_safe_haven}"
+style = [
+  "ruff {args:data_safe_haven}",
+  "black --check --diff {args:data_safe_haven}",
+]
+fmt = [
+  "black {args:data_safe_haven}",
+  "ruff --fix {args:data_safe_haven}",
+  "style",
+]
+all = [
+  "style",
+  "typing",
+]
+
+[tool.black]
+target-version = ["py310", "py311"]
+line-length = 120
+skip-string-normalization = true
+
+[tool.ruff]
+line-length = 120
+select = [
+  "A",
+  "ARG",
+  "B",
+  "C",
+  "DTZ",
+  "E",
+  "EM",
+  "F",
+  "FBT",
+  "I",
+  "ICN",
+  "ISC",
+  "N",
+  "PLC",
+  "PLE",
+  "PLR",
+  "PLW",
+  "Q",
+  "RUF",
+  "S",
+  "T",
+  "TID",
+  "UP",
+  "W",
+  "YTT",
+]
+ignore = [
+  # Allow non-abstract empty methods in abstract base classes
+  "B027",
+  # Allow boolean positional values in function calls, like `dict.get(... True)`
+  "FBT003",
```

Comment:
Do we need this?

Reply:
There are quite a few instances of this rule being broken.

I think there is a good argument to follow this rule. `do_something(..., True)` is not clear and could mean `do_something(..., verbose=True)` or `(do_something(..., dry_run=True)`.
There is a rule to stop _us_ writing functions with position boolean arguments.

Reply:
Removed (at least for now) in f8a5e6f9